packages/space-opera/src/components/model_viewer_snippet/components/parsing.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import {ModelViewerConfig} from '@google/model-viewer-editing-adapter/lib/main';
+import {ArConfigState} from '../../../types';
+
+// A list of accepted attributes in render_model_viewer. Used when parsing for
+// extra attributes.
+const ATTRIBUTE_LIST: string[] = [
+  'src',
+  'auto-rotate',
+  'camera-controls',
+  'environment-image',
+  'skybox-image',
+  'exposure',
+  'poster',
+  'reveal',
+  'shadow-intensity',
+  'shadow-softness',
+  'max-camera-orbit',
+  'min-camera-orbit',
+  'max-field-of-view',
+  'min-field-of-view',
+  'autoplay',
+  'animation-name',
+  'camera-orbit',
+  'camera-target',
+  'field-of-view',
+  'ar',
+  'ar-modes',
+  'ios-src'
+];
+
+// Creating object in this way such that "spread" works.
+// https://open-wc.org/docs/development/lit-helpers/#regular-spread
+export function parseExtraAttributes(snippet: string): string {
+  // Parse snippet and extract attributes from model-viewer
+  const parsedInput = new DOMParser().parseFromString(snippet, 'text/html');
+  const modelViewer = parsedInput.body.getElementsByTagName('model-viewer')[0];
+  let extraAttributes: any = {};
+  const attributes = modelViewer.attributes;
+
+  // Loop through every attribute, only add the attributes to extraAttributes
+  // if they are not a part of config or arConfig
+  for (let i = 0; i < attributes.length; i++) {
+    const name: string = attributes[i].name;
+    const value = attributes[i].value;
+    // if neither config has the key, add it to extra attribute string
+    if (!(ATTRIBUTE_LIST.includes(name))) {
+      if (value.length === 0) {
+        // Boolean attribute is true
+        extraAttributes[`?${name}`] = true;
+      } else {
+        // Normal attribute is set to it's snippet's value
+        extraAttributes[`${name}`] = value;
+      }
+    }
+  }
+  return extraAttributes;
+}
+
+// logic similar to parseSnippet inside of editing adapter.
```

Comment:
Are we still using parseSnippet from editing adapter, or is this the canonical one for space-opera now?

Reply:
yeah, this is the canonical from space opera now